src/merkle/mmr/full.rs

```diff
@@ -0,0 +1,232 @@
+    /// added, this corresponds to the MMR size _prior_ to adding the element. So the 1st alement
+    /// has position 0, the second position 1, and so on.
+    pub fn merkle_path(&self, pos: usize) -> Option<Vec<PathElement<Word>>> {
+        // find the target tree responsible for the MMR position
+        let tree_bit = leaf_to_corresponding_tree(pos, self.forest)?;
+        let forest_target = 1usize << tree_bit;
+
+        // isolate the trees before the target
+        let forest_before = self.forest & high_bitmask(tree_bit + 1);
+        let index_offset = nodes_in_forest(forest_before);
+
+        // update the position from global to the target tree
+        let mut index = nodes_in_forest(forest_target) - 1;
+        let pos = pos - forest_before;
+
+        // collect the Merkle path
+        let tree_depth = (tree_bit + 1) as usize;
+        let mut result = Vec::with_capacity(tree_depth);
+        let mut tree_depth = forest_target >> 1; // the root is not included in the path
+        while tree_depth > 0 {
+            let bit = pos & tree_depth;
+            let right_offset = index - 1;
+            let left_offset = right_offset - nodes_in_forest(tree_depth);
+
+            // Elements to the right have a higher position because they were
+            // added later. Therefore when the bit is true the node's path is
+            // to the right, and its sibling to the left.
+            let sibling = if bit != 0 {
+                index = right_offset;
+                PathElement::LeftSibling(self.nodes[index_offset + left_offset])
+            } else {
+                index = left_offset;
+                PathElement::RightSibling(self.nodes[index_offset + right_offset])
+            };
+
+            tree_depth >>= 1;
+            result.push(sibling);
+        }
+
+        Some(result)
+    }
+
+    // FUNCTIONALITY
+    // ============================================================================================
+
+    /// Adds a new element to the MMR.
+    pub fn add(&mut self, el: Word) {
+        // Note: every node is also a tree of size 1, adding an element to the forest creates a new
+        // rooted-tree of size 1. This may temporarily break the invariant that every tree in the
+        // forest has different sizes, the loop below will eagerly merge trees of same size and
+        // restore the invariant.
+        self.nodes.push(el);
+
+        let mut left_offset = self.nodes.len().saturating_sub(2);
+        let mut right = el;
+        let mut left_tree = 1;
+        while self.forest & left_tree != 0 {
+            right = *Rpo256::hash_elements(&[self.nodes[left_offset], right].concat());
+            self.nodes.push(right);
+
+            left_offset = left_offset.saturating_sub(nodes_in_forest(left_tree));
+            left_tree <<= 1;
+        }
+
+        self.forest += 1;
+    }
+
+    pub fn accumulator(&self) -> Accumulator {
```

Comment:
Is there a reason why the `Accumulator` is a distinct entity as opposed to storing the accumulator data as a member of the MMR and updating it dynamically as leaves are added?

Reply:
I think there are two things here:

1. Right now accumulator data is the same as the tree's roots, and that _is_ stored in `nodes`, so I suppose this was not the point.
2. Maybe you meant why not have `nodes` and `accumulator` at the same time in the `FullMMR` struct? Here we would have these options
     1. add the tree roots to `accumulator` first and then move it to `nodes` as trees are merged. This would be very inefficient, because to make space for the `root`s we would need to move around a large number of nodes, which would result in lots of copies.
     2. alternatively, we could have the space there already, and then the question is just the direction of the copy, we can add to the accumulator first, and then copy to the `nodes` as needed, or the other way around. I went with the other way around, of copying data from `nodes` when we need an accumulator, but they should be pretty much the same.
     3. and there is also the option of just having them both simultaneously, and in that case it would mean that `accumulator` would be called at the end of `add`, but I don't see the benefit since that may be unnecessary work. If the caller needs for every add they can create it, but if they don't need on every add they can skip the cost of creating it.

Reply:
Thank you for the explanation!  As you say I think the current implementation is more efficient.